compiler/msgs.nim

```diff
@@ -66,14 +66,20 @@ when defined(nimpretty):
   proc fileSection*(conf: ConfigRef; fid: FileIndex; a, b: int): string =
     substr(conf.m.fileInfos[fid.int].fullContent, a, b)
 
+proc canonicalCase(path: string): string =
```

Comment:
Please make this a ``var string`` version. All usages call it with a temporary that could be mutated. No allocations then.

Reply:
> All usages call it with a temporary that could be mutated

there's a complication here which is that `canon` can be shallow depending on runtime conditions, and its original value is needed in `newFileInfo(canon...` so you don't end up with ugly paths in error messages like `/users/timothee/git_clone/nim/...` instead of original case `/Users/timothee/git_clone/nim/...`

well, you asked for it ... so I'm introducing lib/std/private/strutils2.nim to do this cleanly.

I could've obviously done things differently, and even forgo this optimization given the fact that `canonicalizePath` calls `expandFile` which in turns calls a filesystem syscall which will mostly likely dwarf any saving here, but the reason I've added `strutils2.nim` is:
* I want to add this at some point so might as well start here
* this is a good staging ground (only private use within compiler + tooling in nim repo) so API can keep improving with real usage and without being bogged down by backward compatibility arguments given that it's private
* this is intended to grow, in particular towards https://github.com/nim-lang/RFCs/issues/191 and several other string API that will benefit from a more performant base
* eventually the dependency can be reversed (eg std/strutils would import strutils2)